source/MaterialXContrib/MaterialXNode/MaterialXData.cpp

```diff
@@ -1,49 +1,93 @@
-#include <MaterialXData.h>
-#include <Util.h>
+#include "MaterialXData.h"
+#include "MaterialXUtil.h"
 
 #include <MaterialXFormat/XmlIo.h>
 #include <MaterialXGenOgsXml/GlslFragmentGenerator.h>
 #include <MaterialXGenShader/Util.h>
 
-MaterialXData::MaterialXData(const std::string& materialXDocumentPath, 
-                             const std::string& elementPath, 
-                             const MaterialX::FileSearchPath& librarySearchPath)
-    : _shaderGenerator(mx::GlslFragmentGenerator::create()),
-      _genContext(_shaderGenerator),
-      _librarySearchPath(librarySearchPath)
+namespace
 {
-    setRenderableElement(materialXDocumentPath, elementPath);
+mx::DocumentPtr createDocument( const std::string& materialXDocumentPath,
+                                const MaterialX::FileSearchPath& librarySearchPath )
+{
+    // Create document
+    mx::DocumentPtr document = mx::createDocument();
+    if (!document)
+    {
+        throw mx::Exception("Failed to create a MaterialX document");
+    }
+
+    // Load libraries
+    static const mx::StringVec libraries = { "stdlib", "pbrlib", "bxdf", "stdlib/genglsl", "pbrlib/genglsl" };
+    MaterialXMaya::loadLibraries(libraries, librarySearchPath, document);
+
+    // Read document contents from disk
+    mx::XmlReadOptions readOptions;
+    readOptions.skipDuplicateElements = true;
+    mx::readFromXmlFile(document, materialXDocumentPath, mx::EMPTY_STRING, &readOptions);
+
+    return document;
+}
 }
 
-bool MaterialXData::setRenderableElement(const std::string& materialXDocument, const std::string& elementPath)
+MaterialXData::MaterialXData(   mx::DocumentPtr document,
+                                const std::string& elementPath, 
+                                const MaterialX::FileSearchPath& librarySearchPath )
+    : _document(document)
+    , _shaderGenerator(mx::GlslFragmentGenerator::create())
+    , _genContext(_shaderGenerator)
 {
-    clearXml();
-    _document = nullptr;
-    _element = nullptr;
+    if (!_document)
+    {
+        throw mx::Exception("No document specified");
+    }
+
+    std::vector<mx::TypedElementPtr> renderableElements;
+    mx::findRenderableElements(_document, renderableElements);
```

Comment:
Checking for renderability right in the constructor because we need to retrieve the list of renderables whether an element has been specified or not and there's less code duplication this way. At the same time, we guarantee that no MaterialXData is created with a non-renderable element.

Reply:
Just a note that finding renderable elements is not that cheap, but should be a big cost.

Reply:
Duly noted. Just to double-check: I believe the cost was there before, I just unified the two code paths.